### PR TITLE
Improve validation message on build spec update

### PR DESCRIPTION
**Pull request by csrwng**
Includes a diff of the build specs (old/new) in the error message details.

### CODE REVIEW

**Comment by csrwng** [comment]
Should help to diagnose https://github.com/fabric8io/openshift-jenkins-sync-plugin/issues/70
@deads2k ptal

cc @jimmidyson @spadgett

File: pkg/build/api/validation/validation.go [hunk 1]
```diff
@@ -36,12 +40,43 @@ func ValidateBuildUpdate(build *buildapi.Build, older *buildapi.Build) field.Err
...
+	}
+	olderJSON, err := runtime.Encode(codec, olderObj)
+	if err != nil {
+		return "", fmt.Errorf("error encoding older: %v", err)
+	}
+	versioned, err := kapi.Scheme.ConvertToVersion(newerObj, version.String())
+	if err != nil {
+		return "", fmt.Errorf("error converting newer to versioned: %v", err)
+	}
+	patch, err := strategicpatch.CreateTwoWayMergePatch(olderJSON, newerJSON, versioned)
```
Comment (deads2k) [inline pkg/build/api/validation/validation.go]: can `versioned` just be an empty build?  I'm pretty sure it just uses it navigate metadata for strategic merge patches.

**Comment by deads2k** [comment]
I think we may want to refactor it more generally, but this is a good start.

**Comment by csrwng** [comment]
comment addressed

File: pkg/build/api/validation/validation_test.go [hunk 2]
```diff
@@ -2320,3 +2322,79 @@ func TestValidatePostCommit(t *testing.T) {
...
+				},
+			},
+			newer: buildapi.BuildSpec{
+				CommonSpec: buildapi.CommonSpec{
+					Source: buildapi.BuildSource{
+						ContextDir: "context-dir",
+					},
+				},
+			},
+			expected: `{"Spec":{"Source":{"ContextDir":"context-dir"}}}`,
```
Comment (deads2k) [inline pkg/build/api/validation/validation_test.go]: you shouldn't be getting caps.  Convert to external first.

Comment (csrwng) [inline pkg/build/api/validation/validation_test.go]: which external? I didn't want to assume 'v1' is that ok?

File: pkg/build/api/validation/validation.go [hunk 3]
```diff
@@ -36,12 +40,38 @@ func ValidateBuildUpdate(build *buildapi.Build, older *buildapi.Build) field.Err
...
 
 	return allErrs
 }
 
+func diffBuildSpec(newer buildapi.BuildSpec, older buildapi.BuildSpec) (string, error) {
+	codec := kapi.Codecs.LegacyCodec(buildapi.SchemeGroupVersion)
+	newerObj := &buildapi.Build{Spec: newer}
+	olderObj := &buildapi.Build{Spec: older}
+
+	newerJSON, err := runtime.Encode(codec, newerObj)
```
Comment (deads2k) [inline pkg/build/api/validation/validation.go]: Specifying the version here ought to encode you to a version.  You'll need need the obj you pass to `CreateTwoWayMergePatch` to match version-wise.

File: pkg/build/api/validation/validation_test.go (hunk 2 above)
Comment (deads2k) [inline pkg/build/api/validation/validation_test.go]: > which external? I didn't want to assume 'v1' is that ok?

Yeah, v1 is closer than anything else right now.

**Comment by csrwng** [comment]
using v1 now

**Comment by csrwng** [comment]
btw, I'd considered using v1 before, but wasn't sure that it was proper to reference a specific version directly.
maybe we need a build/api/latest ?

**Comment by deads2k** [comment]
> maybe we need a build/api/latest ?

It's not really latest.  Validation currently chooses an API version that its reported tags (the paths) are going to match.  You should probably match whatever that is.

**Comment by deads2k** [comment]
lgtm [merge]

**Comment by csrwng** [comment]
[merge]

**Comment by csrwng** [comment]
[merge]

**Comment by deads2k** [comment]
[merge]

**Comment by csrwng** [comment]
[merge]

**Pull request merged**